### DISCUSSION TITLE
How to filter rows based on value in jsonb array?

### DISCUSSION THREAD

**Discussion by radim10** [supabase/supabase#7373] [Questions]
I have tasks table with column subtasks. The subtasks column is jsonb array, for example: `[{text:"some text", priority:1, id:934873}]`. With the js library, I want to filter all tasks where the jsonb array contains id (not all object keys and values) which is equal to some value. I have tried this solution: `.contains('subtasks:subtasks->>subtasks', ['subtasks->id: 66753'])`, but it does not work. Any solutions?

**Comment by GaryAustin1** [comment]
This seems similar:  https://github.com/supabase/supabase/discussions/2102

**Reply by radim10** [reply]
Yes, it is similar. But the solution there does not work for me, it is returning an empty array.

**Comment by silentworks** [comment]
Can you provide the full query you are running because the partial you provided here is a little hard to work with? We don't know what `subtasks:subtasks` is.

**Reply by radim10** [reply]
The subtasks column has this TS type: `Array<{text:string; priority: number; id:number}>`. I have tried this solution [Stack Overflow](https://stackoverflow.com/questions/68406176/supabase-json-query-javascriptl), so in my case: `  const { data, error } = await supabase       .from('tasks').select('id, subtasks').contains('subtasks->>subtask', ['subtask->id: 3423'])` but this gives me an error: _operator does not exist: text @> unknown_. I also tried this: `    const { data, error } = await supabase.from('tasks').select('id, subtasks').eq('subtasks->id', 3423)` which does not give me an error but instead it always returns an empty array. What am I missing here?

**Comment by lerayj** [comment]
@radim10 Did you manage to find a solution ? I'm looking for the same thing, and having the same output (empty array).
Would love to hear from you ! 

EDIT:
Actually it's easier to just change the column type to text[]. Then no more hassle

**Comment by radim10** [comment]
@lerayj I asked on Reddit too and one answer by @kiwicopple suggest querying JSONB array like this:
`const { data, error } = await supabase.from('users').select().eq('subtasks -> id', 934873)`. Or you can create pure SQL function and call it with RPC.

**Comment by ThomasBnf** [comment]
Ok i found why it was not working.
My Jsonb was define as an **array**, **_avoid it._**

Secondly i would like to filter values by different operator so i created a function

This function is now dynamic, and i can call it in my code with supabase RPC

This is my function : 

CREATE OR REPLACE FUNCTION get_maintenances(
  search_column text,
  search_operator text,
  search_value text
) RETURNS SETOF maintenances AS $$
BEGIN
  RETURN QUERY EXECUTE format(
    'SELECT m.*
     FROM maintenances m
     CROSS JOIN jsonb_array_elements(m.values) as fields
     WHERE fields->>%L %s %L',
     search_column, search_operator, search_value
  );
END;
$$ LANGUAGE plpgsql;


Replace "maintenances" by your table name and m."values" by your column name and name it as you wish.

now you can execute it in your code

const fetchMaintenances = async () => {
    const {data, error} = await supabase.rpc("get_maintenances", {
      search_column: "value",
      search_operator: "ilike",
      search_value: "f%"
    })
    return data
  }

Enjoy !

**Comment by ahmadawais** [comment]
Anyone still looking at this, make sure you read this:

> Postgres support a range of [JSON functions and operators](https://www.postgresql.org/docs/current/functions-json.html). For example, the -> operator returns values as jsonb data. If you want the data returned as text, use the ->> operator. — [Source](https://supabase.com/docs/guides/database/json?database-method=js&language=js)

So for comparison you should use `->>` which will return text value instead of jsonb.

**Reply by santhonydo** [reply]
omg. was struggling with this for a few hours. thanks this solved it.

**Reply by vidicunt** [reply]
> omg. was struggling with this for a few hours. thanks this solved it.

Could you please share your solution?

**Comment by vidicunt** [comment]
@radim10 did you manage to find a solution?
